components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/utils/OpenAPICodegenUtils.java

```diff
@@ -354,26 +380,56 @@ private static EndpointListRouteDTO extractEndpointFromOpenAPI(Object endpointEx
      *
      * @param openAPI         {@link OpenAPI} object
      * @param openApiFilePath OpenAPI definition file
+     * @param openAPIVersion OpenAPI version
      */
-    private static void validateBasepath(OpenAPI openAPI, String openApiFilePath) {
+    private static void validateBasePath(OpenAPI openAPI, String openApiFilePath, String openAPIVersion) {
+        List<Server> servers = openAPI.getServers();
         Map<String, Object> openAPIExtensions = openAPI.getExtensions();
-        if (openAPIExtensions == null) {
-            throw new CLIRuntimeException(" At least '" + OpenAPIConstants.BASEPATH + "' property and '"
-                    + OpenAPIConstants.PRODUCTION_ENDPOINTS + "' property should present in the open API "
-                    + "definition: '" + openApiFilePath + "'.");
+        String basePath = null;
+        if (openAPIExtensions != null && openAPIExtensions.get(OpenAPIConstants.BASEPATH) != null) {
+            basePath = (String) openAPIExtensions.get(OpenAPIConstants.BASEPATH);
+        } else if (servers != null) {
+            //We  need to check only the first server url to derive the base path. Each and every server url
+            // should contains the same base path.
```

Comment:
There is a confusion here. For each resource and path, we consider the servers object to be the backend. But when it is in the API level, we take that to define the path exposed from the gateway as well. WDYT?

Reply:
IMO, if the user does not provide the basepath via the extension, we should not expose the actual backend's basepath.

Reply:
Here, we add the server url base path as the micro gateway base path, and hence get added to every path level operation as well.
We should not enforce x-wso2-basePath , so thats why we need to take the base path from the server url.

Reply:
When there are path level/ operation level endpoints, it will not be an issue. But if the developer is relying on API level endpoints, it is not appropriate to expose the basepath of his backend. IMO, that is a piece of information we should not  expose to the users of API.

Reply:
I don't think exposing base path as a threat or security concern, It is the backend URL(IP/host and port) that we should not expose to the outside